### PR TITLE
packer console for HCL2

### PR DESCRIPTION
**Pull request by azr**
* [x] `packer console` for hcl2 & json
* [x] reintroduce `packer console` when no argument is passed
  *   [x] add `--hcl2` option for that
* [x] add unit tests
* [x] test piped commands: `echo "var.something" | packer console --hcl2`

### CODE REVIEW

File: packer/run_interfaces.go [hunk 1]
```diff
@@ -17,6 +17,19 @@ type BuildGetter interface {
...
 
+type Evaluator interface {
+	// EvaluateExpression is meant to be used in the `packer console` command.
+	// It parses the input string and returns what needs to be displayed. In
+	// case of an error the error should be displayed.
+	EvaluateExpression(expr string) (output string, exit bool, diags hcl.Diagnostics)
+}
+
+// The packer.Handler handles all Packer things.
+type Handler interface {
```
Comment (azr) [inline packer/run_interfaces.go:28]: Could/should this be a bit better ?

**Comment by azr** [comment]
Okay this is ready for review; I added means to test commands from https://github.com/golang/go/blob/master/src/os/exec/exec_test.go#L72 and this is really awesome :D

File: command/build.go [hunk 2]
```diff
@@ -88,15 +87,15 @@ func writeDiags(ui packer.Ui, files map[string]*hcl.File, diags hcl.Diagnostics)
...
+	cfgType, err := cla.GetConfigType()
 	if err != nil {
-		m.Ui.Error(fmt.Sprintf("could not tell config type: %s", err))
+		m.Ui.Error(fmt.Sprintf("%q: %s", cla.Path, err))
 		return nil, 1
 	}
 
 	switch cfgType {
-	case "hcl":
+	case ConfigTypeHCL2:
```
Comment (SwampDragons) [inline command/build.go:98]: good call changing this to a constant.

File: packer/run_interfaces.go (hunk 1 above)
Comment (SwampDragons) [reply packer/run_interfaces.go:28]: I think it's fine.

File: command/console.go [hunk 3]
```diff
@@ -128,24 +100,26 @@ func (*ConsoleCommand) AutocompleteFlags() complete.Flags {
...
+func (c *ConsoleCommand) modePiped(cfg packer.Evaluator) int {
 	var lastResult string
 	scanner := bufio.NewScanner(wrappedstreams.Stdin())
+	ret := 0
 	for scanner.Scan() {
-		result, err := session.Handle(strings.TrimSpace(scanner.Text()))
-		if err != nil {
-			return 0
+		result, _, diags := cfg.EvaluateExpression(strings.TrimSpace(scanner.Text()))
+		if len(diags) > 1 {
```
Comment (nywilken) [inline command/console.go:109]: In the case that we only have one item in diags should we print it? Or should this be if `len(diags) > 0`?

File: command/cli.go [hunk 4]
```diff
@@ -2,40 +2,51 @@ package command
...
-	"fmt"
 	"strings"
 
 	"github.com/hashicorp/packer/helper/enumflag"
 	kvflag "github.com/hashicorp/packer/helper/flag-kv"
 	sliceflag "github.com/hashicorp/packer/helper/flag-slice"
 )
 
+//go:generate enumer -type configType -trimprefix ConfigType -transform snake
+type configType int
```
[the remark this answers is not in the document]
Comment (azr) [reply command/cli.go:13]: me too ! 🙂

File: command/console.go (hunk 3 above)
Comment (azr) [reply command/console.go:109]: Yes totally !

Comment (azr) [reply command/console.go:109]:
```suggestion
		if len(diags) > 0 {
```

File: hcl2template/repl/format.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	"sort"
+	"strconv"
+	"strings"
+)
+
+// FormatResult formats the given result value for human-readable output.
+//
+// The value must currently be a string, list, map, and any nested values
+// with those same types.
+func FormatResult(value interface{}) string {
```
Comment (sylviamoss) [inline hcl2template/repl/format.go:16]: interesting! this can be useful for other scopes in the future :D

Comment (azr) [reply hcl2template/repl/format.go:16]: yes ! :D

**Comment by azr** [comment]
Okay I think this is good enough to merge and merging this to unblock Wilken !

**Pull request merged**